yew-functional/src/lib.rs

```diff
@@ -173,204 +169,76 @@ where
+                let mut hook = internal_hook_state.borrow_mut();
+                let hook: &mut T = hook.downcast_mut().expect("Wrong type");
+                cb(hook)
+            }),
+            post_render,
+        );
     }
-    impl<T: FnOnce() + 'static> Hook for UseEffectState<T> {
-        fn tear_down(&mut self) {
-            if let Some(destructor) = self.destructor.take() {
-                destructor()
-            }
-        }
-    }
-
-    let callback = Box::new(callback);
-
-    use_hook(
-        |_: &mut UseEffectState<Destructor>, hook_callback| {
-            hook_callback(
-                move |state: &mut UseEffectState<Destructor>| {
-                    if let Some(de) = state.destructor.take() {
-                        de();
-                    }
-                    let new_destructor = callback();
-                    state.destructor.replace(Box::new(new_destructor));
-                    false
-                },
-                true, // run post render
-            );
-        },
-        || UseEffectState { destructor: None },
-    );
 }
 
-pub fn use_effect_with_deps<F, Destructor, Dependents>(callback: F, deps: Dependents)
-where
-    F: FnOnce(&Dependents) -> Destructor + 'static,
-    Destructor: FnOnce() + 'static,
-    Dependents: PartialEq + 'static,
-{
-    struct UseEffectState<Dependents, Destructor> {
-        deps: Rc<Dependents>,
-        destructor: Option<Box<Destructor>>,
-    }
-    impl<Dependents, Destructor: FnOnce() + 'static> Hook for UseEffectState<Dependents, Destructor> {
-        fn tear_down(&mut self) {
-            if let Some(destructor) = self.destructor.take() {
-                destructor()
-            }
-        }
-    }
-
-    let deps = Rc::new(deps);
-    let deps_c = deps.clone();
-
-    use_hook(
-        move |_state: &mut UseEffectState<Dependents, Destructor>, hook_callback| {
-            hook_callback(
-                move |state: &mut UseEffectState<Dependents, Destructor>| {
-                    if state.deps != deps {
-                        if let Some(de) = state.destructor.take() {
-                            de();
-                        }
-                        let new_destructor = callback(deps.borrow());
-                        state.deps = deps;
-                        state.destructor.replace(Box::new(new_destructor));
-                    } else if state.destructor.is_none() {
-                        state
-                            .destructor
-                            .replace(Box::new(callback(state.deps.borrow())));
-                    }
-                    false
-                },
-                true, // run post render
-            );
-        },
-        || UseEffectState {
-            deps: deps_c,
-            destructor: None,
-        },
-    );
+pub trait Hook {
+    type Output;
+    type Args;
+    fn tear_down(&mut self) {}
+    fn runner(&mut self, args: Self::Args, updater: HookUpdater) -> Self::Output;
```

Comment:
Shouldn't this be called "run" or is there a reason you went for "runner"?

Reply:
Less relevant now with the explicit lifecycle without the "Hook" trait.